libmscore/rendermidi.cpp

```diff
@@ -628,61 +628,150 @@ void Score::renderStaff(EventMap* events, Staff* staff)
+      {
+      // TODO don't have this code multiple times - maybe have one method with switch-case?
+      int tickOffset = rs->utick - rs->tick;
+      int utick1 = rs->utick;
+      int tick1 = repeatList()->utick2tick(utick1);
+      int tick2 = tick1 + rs->len;
+
+      midiChanValMap *channelMap = &CCEvents.at(s->ccNumber());
+
+      int idx = s->staff()->channel(s->tick(), 0);
+      int channel = s->part()->instrument(s->tick())->channel(idx)->channel;
+      if (channelMap->find(channel) == channelMap->end())
+            channelMap->insert({channel, midiValList()});
+
+      // Borders in which modulation takes place
+      int startTick = -1;
+      int endTick   = -1;
+      int spannerOffset = 0; // Ticks offset sice spanner take on
+
+      if (s->tick() >= tick1 && s->tick() < tick2)
+            startTick = s->tick();
+
+      if (s->tick2() <= tick2 && s->tick2() > tick1)
+            endTick = s->tick2();
+
+      // Spanner is outside of this RepeatSegment
+      if (startTick == -1 && endTick == -1)
+            return;
+
+      // Spanner starts before this RepeatSegment
+      if (startTick == -1) {
+            startTick = tick1;
+            spannerOffset = tick1 - startTick;
+            }
+
+      // Spanner ends outside this RepeatSegment
+      if (endTick == -1)
+            endTick = tick2;
+
+      int ticksPerMidiInc = s->ticks() / (s->ccEnd() - s->ccStart());
+      int curMidiVal = (spannerOffset * ticksPerMidiInc) + s->ccStart();
+
+      for(int pos = startTick; pos < endTick; pos += ticksPerMidiInc) {
+            channelMap->at(channel).push_back(midiTimeVal(pos + tickOffset, curMidiVal));
+            curMidiVal++;
+            }
+
+      // make sure we have an end value
+      channelMap->at(channel).push_back(midiTimeVal(endTick, (endTick*ticksPerMidiInc) + s->ccStart()));
+
+      }
+
+//---------------------------------------------------------
+//   renderSwitchSpanners
+//---------------------------------------------------------
+
+void Score::renderSpannerMidi(EventMap* events, midiCCMap CCEvents) {
+      for (auto CCEvent : CCEvents) {
+            int cc = CCEvent.first;
+            midiChanValMap channelEvents = CCEvent.second;
+            for (const auto& channelEvent : channelEvents) {
+                  int channel = channelEvent.first;
+                  for (const auto& pe : channelEvent.second) {
+                        NPlayEvent event;
+                        event = NPlayEvent(ME_CONTROLLER, channel, cc, pe.second);
+                        events->insert(std::pair<int,NPlayEvent>(pe.first, event));
+                        }
+                  }
+            }
+      }
+
+//---------------------------------------------------------
 //   renderSpanners
 //---------------------------------------------------------
 
 void Score::renderSpanners(EventMap* events, int staffIdx)
       {
+      midiCCMap CCEvents = midiCCMap();
       foreach (const RepeatSegment* rs, *repeatList()) {
```

Comment:
while touching the code, this is a good opporunity to change to C++11 `for (... : ...)`